wgpu/tests/shader/zero_init_workgroup_mem.rs

```diff
@@ -0,0 +1,150 @@
+use wgpu::{
+    include_wgsl, BindGroupDescriptor, BindGroupEntry, BindGroupLayoutDescriptor,
+    BindGroupLayoutEntry, BindingType, BufferBindingType, BufferDescriptor, BufferUsages,
+    CommandEncoderDescriptor, ComputePassDescriptor, ComputePipelineDescriptor, DownlevelFlags,
+    Limits, Maintain, MapMode, PipelineLayoutDescriptor, ShaderStages,
+};
+
+use crate::common::{initialize_test, TestParameters, TestingContext};
+
+#[test]
+fn zero_init_workgroup_mem() {
+    initialize_test(
+        TestParameters::default()
+            .downlevel_flags(DownlevelFlags::COMPUTE_SHADERS)
+            .limits(Limits::downlevel_defaults()),
+        zero_init_workgroup_mem_impl,
+    );
+}
+
+/// Increases iterations and writes random data to workgroup memory before reading it each iteration.
+const TRY_TO_FAIL: bool = false;
```

Comment:
I feel like this should just always be on

Reply:
This is more for testing for failures specifically since I noticed that I was getting quite a few false positives when testing for failures. Turning it on still doesn't guarantee that it will fail since some drivers seem to zero initialize workgroup memory. Maybe the name doesn't really help convey this.

Thoughts?

Reply:
Fundamentally what we care about is that it gets zero initialized somehow either by us or the driver.

So if we try our best to make it fail, then it succeeds, we know we're getting what we need done. Without trying our best to fail we might just be getting extra lucky.

Reply:
Hmm, I see what you mean but it will increase the running time of the test quite a bit. Maybe we should lower the iterations to 10 or 20? I just picked 100 randomly for my manual testing.

Reply:
Would dispatching 64x64 read workgroups (and having them all write their value to the buffer we check to be zero) make it such that we hit the non-zero-init the first time? Shouldn't increase execution time a lot but should make it more reproducable?

Reply:
Actually now that I look at it, there's no actual synchronization dependency between the two workgroups - that might also exacerbate the problem (if they both had the bg binding)

Reply:
I used to have the workgroup size for the read pipeline be 64x64 but FXC was complaining about the race condition and I just set it to 1.

Reply:
I think there is no way to 100% make sure this test passes because we are initializing the memory since the driver is free to do so (and some do; I encountered it on my dedicated gpu on the DX12 backend) and cause it to succeed.

Reply:
What I mean for reading is a workgroup size of 1, but doing a 64x64 dispatch, writing out the contents of every workgroup's memory separately. Then compare all of them to see if they're all zero as expected.

We can't make sure 100% that the test fails every time if our zero-init breaks, but we can raise that chance has high as possible, which is what spamming values into workgroup memory does. We want to try to write to and read from every workgroup's memory.

When our zero-init is working, it should always pass. It would false positive on a zero-init-ing driver/gpu, but should be as close to reliable as it can be on a non-zero-init-ing gpu because of how many workgroups we're spamming.

Hope that makes sense

Reply:
Ah, I see what you mean. Increasing the "read" dispatch makes sense. Will update the test tomorrow.

Regarding the nr of iterations, should we still do like 10 or do you think having a read dispatch of 64x64x64 be enough?

Reply:
The fewest amount of iterations that consistently results in failures on drivers without zero init.